Test failure:

[ERROR] Tests run: 25, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.521 s <<< FAILURE! - in com.yahoo.config.application.api.DeploymentSpecTest
[ERROR] emptyNotifications(com.yahoo.config.application.api.DeploymentSpecTest)  Time elapsed: 0.173 s  <<< FAILURE!
java.lang.AssertionError: expected:<com.yahoo.config.application.api.Notifications@402f32ff> but was:<com.yahoo.config.application.api.Notifications@573f2bb1>
	at com.yahoo.config.application.api.DeploymentSpecTest.emptyNotifications(DeploymentSpecTest.java:454)

--- a/config-model-api/src/main/java/com/yahoo/config/application/api/Notifications.java
+++ b/config-model-api/src/main/java/com/yahoo/config/application/api/Notifications.java
@@ -11,7 +11,6 @@ import java.util.Map;
 import java.util.Set;
 
 import static java.util.Collections.emptyList;
-import static java.util.stream.Collectors.toMap;
 
 /**
  * Configuration of notifications for deployment jobs.
@@ -45,7 +44,8 @@ public class Notifications {
      * @return The Notifications as specified.
      */
     public static Notifications of(Map<When, List<String>> emailAddressesByWhen, Map<When, List<Role>> emailRolesByWhen) {
-        if (emailAddressesByWhen.isEmpty() && emailRolesByWhen.isEmpty())
+        if (   emailAddressesByWhen.values().stream().allMatch(List::isEmpty)
+            && emailRolesByWhen.values().stream().allMatch(List::isEmpty))
             return none;
 
         ImmutableMap.Builder<When, List<String>> emailAddresses = ImmutableMap.builder();

--- a/config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java
+++ b/config-model-api/src/main/java/com/yahoo/config/application/api/xml/DeploymentSpecXmlReader.java
@@ -128,7 +128,7 @@ public class DeploymentSpecXmlReader {
         Map<When, List<Role>> emailRoles = new HashMap<>();
         for (When when : When.values()) {
             emailAddresses.put(when, new ArrayList<>());
-            emailRoles.putIfAbsent(when, new ArrayList<>());
+            emailRoles.put(when, new ArrayList<>());
         }
 
         for (Element emailElement : XML.getChildren(notificationsElement, "email")) {

--- a/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/InternalStepRunner.java
+++ b/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/InternalStepRunner.java
@@ -46,8 +46,6 @@ import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;

